### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Roy-Hu**

### CODE REVIEW

**Comment by khirono** [comment]
Would you delete a binary file?

cmd/gogtp5g-link/gogtp5g-link

File: cmd.go [hunk 1]
```diff
@@ -22,5 +22,6 @@ const (
 	CMD_GET_URR
 	CMD_GET_BAR
 
+	CMD_GET_REPORT
```
Comment (khirono) [inline cmd.go]: New definition should be after CMD_GET_VERSION for compatability with older version of free5GC's UPF.

Comment (Roy-Hu) [reply cmd.go]: Thanks for notifying.  I have made some changes based on these comments.

File: attr_report.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	UR_VOLUME_MEASUREMENT_UVOL
+	UR_VOLUME_MEASUREMENT_DVOL
+
+	UR_VOLUME_MEASUREMENT_TOPACKET
+	UR_VOLUME_MEASUREMENT_UPACKET
+	UR_VOLUME_MEASUREMENT_DPACKET
+)
+
+type USAReport struct {
+	URRID          uint32
```
Comment (khirono) [inline attr_report.go:40]: How does UPF notify usage report to an associated session?
I think USAReport needs to have the SEID field.

File: attr_report.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	URSEQN         uint32
+	USARTrigger    uint32
+	VolMeasurement VolumeMeasurement
+	QueryUrrRef    uint32
+}
```
Comment (khirono) [inline attr_report.go:47]: USAReport needs to contains Start Time and End Time fields.

> Start Time
> This IE shall be present, except if the Usage Report Trigger indicates 'Start of Traffic', 'Stop of Traffic' or 'MAC Addresses Reporting'.
> When present, this IE shall provide the timestamp when the collection of the information in this report was started.

> End Time
> This IE shall be present, except if the Usage Report Trigger indicates 'Start of Traffic', 'Stop of Traffic' or ' MAC Addresses Reporting'.
> When present, this IE shall provide the timestamp when the collection of the information in this report was generated.

File: attr_report.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+func DecodeVolumeMeasurement(b []byte) (VolumeMeasurement, error) {
+	var VolMeasurement VolumeMeasurement
+	for len(b) > 0 {
+		hdr, n, err := nl.DecodeAttrHdr(b)
+		if err != nil {
+			return VolMeasurement, err
+		}
+		switch hdr.MaskedType() {
```
Comment (khirono) [inline attr_report.go:90]: Flag is computable without get from gtp5g.

```go
const (
	TOVOL uint8 = 1 << iota
	ULVOL
	DLVOL
	TONOP
	ULNOP
	DLNOP
)
...snip...
		switch hdr.MaskedType() {
		case UR_VOLUME_MEASUREMENT_TOVOL:
			v := native.Uint64(b[n:])
			VolMeasurement.TotalVolume = v
			VolMeasurement.Flag |= TOVOL
		case UR_VOLUME_MEASUREMENT_UVOL:
			v := native.Uint64(b[n:])
			VolMeasurement.UplinkVolume = v
			VolMeasurement.Flag |= ULVOL
		case UR_VOLUME_MEASUREMENT_DVOL:
			v := native.Uint64(b[n:])
			VolMeasurement.DownlinkVolume = v
			VolMeasurement.Flag |= DLVOL
		case UR_VOLUME_MEASUREMENT_TOPACKET:
			v := native.Uint64(b[n:])
			VolMeasurement.TotalPktNum = v
			VolMeasurement.Flag |= TONOP
		case UR_VOLUME_MEASUREMENT_UPACKET:
			v := native.Uint64(b[n:])
			VolMeasurement.UplinkPktNum = v
			VolMeasurement.Flag |= ULNOP
		case UR_VOLUME_MEASUREMENT_DPACKET:
			v := native.Uint64(b[n:])
			VolMeasurement.DownlinkPktNum = v
			VolMeasurement.Flag |= DLNOP
		}
```

File: attr_report.go (hunk 2 above)
Comment (Roy-Hu) [reply attr_report.go:40]: Currently, gtp5g will use unix_sock_send() to send a message containing SEID and reports to UPF. The UPF can then decode the SEID and find the associate session. Do I need to integrate the SEID into the report struct?

Comment (khirono) [reply attr_report.go:40]: That's fine. I confirmed that SEID is notified besides this structure.

File: cmd/gogtp5g-tunnel/cmd_urr.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+func CmdAddURR(args []string) error {
+	if len(args) < 2 {
+		return errors.New("too few parameter")
+	}
+	ifname := args[0]
+	oid, err := ParseOID(args[1])
+	if err != nil {
+		return err
+	}
+	attrs, err := ParseQEROptions(args[2:])
```
Comment (khirono) [inline cmd/gogtp5g-tunnel/cmd_urr.go]: typo?
```
attrs, err := ParseURROptions(args[2:])
```

File: cmd/gogtp5g-tunnel/cmd_urr.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+	if err != nil {
+		return err
+	}
+
+	link, err := gtp5gnl.GetLink(ifname)
+	if err != nil {
+		return err
+	}
+
+	return gtp5gnl.UpdateQEROID(c, link, oid, attrs)
```
Comment (khirono) [inline cmd/gogtp5g-tunnel/cmd_urr.go]: typo?
```
return gtp5gnl.UpdateURROID(c, link, oid, attrs)
```

File: cmd/gogtp5g-tunnel/cmd_urr.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+func CmdModURR(args []string) error {
+	if len(args) < 2 {
+		return errors.New("too few parameter")
+	}
+	ifname := args[0]
+	oid, err := ParseOID(args[1])
+	if err != nil {
+		return err
+	}
+	attrs, err := ParseQEROptions(args[2:])
```
Comment (khirono) [inline cmd/gogtp5g-tunnel/cmd_urr.go]: typo?
```
attrs, err := ParseURROptions(args[2:])
```

File: cmd/gogtp5g-tunnel/cmd_urr.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+	if err != nil {
+		return err
+	}
+
+	link, err := gtp5gnl.GetLink(ifname)
+	if err != nil {
+		return err
+	}
+
+	return gtp5gnl.RemoveQEROID(c, link, oid)
```
Comment (khirono) [inline cmd/gogtp5g-tunnel/cmd_urr.go]: typo?
```
return gtp5gnl.RemoveURROID(c, link, oid)
```

File: cmd/gogtp5g-tunnel/cmd_urr.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+	if err != nil {
+		return err
+	}
+
+	link, err := gtp5gnl.GetLink(ifname)
+	if err != nil {
+		return err
+	}
+
+	qer, err := gtp5gnl.GetQEROID(c, link, oid)
```
Comment (khirono) [inline cmd/gogtp5g-tunnel/cmd_urr.go]: typo?
```
urr, err := gtp5gnl.GetURROID(c, link, oid)
```

File: cmd/gogtp5g-tunnel/cmd_urr.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+		return err
+	}
+	defer conn.Close()
+
+	c, err := gtp5gnl.NewClient(conn, mux)
+	if err != nil {
+		return err
+	}
+
+	qers, err := gtp5gnl.GetQERAll(c)
```
Comment (khirono) [inline cmd/gogtp5g-tunnel/cmd_urr.go]: typo?
```
urrs, err := gtp5gnl.GetURRAll(c)
```

File: attr_report.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+			VolMeasurement.DownlinkPktNum = v
+			VolMeasurement.Flag |= DLNOP
+		}
+
+		b = b[hdr.Len.Align():]
+	}
+	return VolMeasurement, nil
+}
+
+func DecodeReport(b []byte) ([]USAReport, error) {
```
Comment (khirono) [inline attr_report.go]: DecodeReport should only return one USAReport.
DecodeReports that decode multiple usage reports can be defined by using one of the following methods:

A) Add number of usage reports in a netlink message
B) Repeatedly call DecodeReport until length of a netlink message equals zero

And then "DecodeUSAReport" is better  than "DecodeReport".

File: cmd/gogtp5g-tunnel/cmd_urr.go [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+	attrs, err := ParseQEROptions(args[2:])
+	if err != nil {
+		return err
+	}
+
+	mux, err := nl.NewMux()
+	if err != nil {
+		return err
+	}
+	defer mux.Close()
```
Comment (khirono) [inline cmd/gogtp5g-tunnel/cmd_urr.go]: Please check new source code because it was modified at #2.

Comment (khirono) [reply cmd/gogtp5g-tunnel/cmd_urr.go]: There is still a problem.

```go
        var wg sync.WaitGroup
        mux, err := nl.NewMux()
        if err != nil {
                return err
        }
        defer func() {
                mux.Close()
                wg.Wait()
        }()
        wg.Add(1)
        go func() {
                mux.Serve()
                wg.Done()
        }()
```

**Pull request merged**